// _@antv_g-canvas@0.4.15@@antv/g-canvas/lib/shape/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var base_1 = require("./base");
Object.defineProperty(exports, "Base", { enumerable: true, get: function () { return base_1.default; } });
var circle_1 = require("./circle");
Object.defineProperty(exports, "Circle", { enumerable: true, get: function () { return circle_1.default; } });
var ellipse_1 = require("./ellipse");
Object.defineProperty(exports, "Ellipse", { enumerable: true, get: function () { return ellipse_1.default; } });
var image_1 = require("./image");
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return image_1.default; } });
var line_1 = require("./line");
Object.defineProperty(exports, "Line", { enumerable: true, get: function () { return line_1.default; } });
var marker_1 = require("./marker");
Object.defineProperty(exports, "Marker", { enumerable: true, get: function () { return marker_1.default; } });
var path_1 = require("./path");
Object.defineProperty(exports, "Path", { enumerable: true, get: function () { return path_1.default; } });
var polygon_1 = require("./polygon");
Object.defineProperty(exports, "Polygon", { enumerable: true, get: function () { return polygon_1.default; } });
var polyline_1 = require("./polyline");
Object.defineProperty(exports, "Polyline", { enumerable: true, get: function () { return polyline_1.default; } });
var rect_1 = require("./rect");
Object.defineProperty(exports, "Rect", { enumerable: true, get: function () { return rect_1.default; } });
var text_1 = require("./text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return text_1.default; } });
