// _@antv_g6@3.5.2@@antv/g6/lib/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.G6Event = void 0;
// Behavior type file
var G6Event;
exports.G6Event = G6Event;

(function (G6Event) {
  G6Event["CLICK"] = "click";
  G6Event["MOUSEDOWN"] = "mousedown";
  G6Event["MOUDEUP"] = "mouseup";
  G6Event["DBLCLICK"] = "dblclick";
  G6Event["CONTEXTMENU"] = "contextmenu";
  G6Event["MOUSEENTER"] = "mouseenter";
  G6Event["MOUSEOUT"] = "mouseout";
  G6Event["MOUSEOVER"] = "mouseover";
  G6Event["MOUSEMOVE"] = "mousemove";
  G6Event["MOUSELEAVE"] = "mouseleave";
  G6Event["DRAGSTART"] = "dragstart";
  G6Event["DRAGEND"] = "dragend";
  G6Event["DRAG"] = "drag";
  G6Event["DRAGENTER"] = "dragenter";
  G6Event["DRAGLEAVE"] = "dragleave";
  G6Event["DRAGOVER"] = "dragover";
  G6Event["DRAGOUT"] = "dragout";
  G6Event["DDROP"] = "drop";
  G6Event["KEYUP"] = "keyup";
  G6Event["KEYDOWN"] = "keydown";
  G6Event["WHEEL"] = "wheel";
  G6Event["FOCUS"] = "focus";
  G6Event["NODE_CLICK"] = "node:click";
  G6Event["NODE_CONTEXTMENU"] = "node:contextmenu";
  G6Event["NODE_DBLCLICK"] = "node:dblclick";
  G6Event["NODE_DRAGSTART"] = "node:dragstart";
  G6Event["NODE_DRAG"] = "node:drag";
  G6Event["NODE_DRAGEND"] = "node:dragend";
  G6Event["NODE_MOUSEENTER"] = "node:mouseenter";
  G6Event["NODE_MOUSELEAVE"] = "node:mouseleave";
  G6Event["NODE_MOUSEMOVE"] = "node:mousemove";
  G6Event["NODE_DROP"] = "node:drop";
  G6Event["NODE_DRAGENTER"] = "node:dragenter";
  G6Event["NODE_DRAGLEAVE"] = "node:dragleave";
  G6Event["EDGE_CLICK"] = "edge:click";
  G6Event["EDGE_CONTEXTMENU"] = "edge:contextmenu";
  G6Event["EDGE_DBLCLICK"] = "edge:dblclick";
  G6Event["EDGE_MOUSEENTER"] = "edge:mouseenter";
  G6Event["EDGE_MOUSELEAVE"] = "edge:mouseleave";
  G6Event["EDGE_MOUSEMOVE"] = "edge:mousemove";
  G6Event["CANVAS_MOUSEDOWN"] = "canvas:mousedown";
  G6Event["CANVAS_MOUSEMOVE"] = "canvas:mousemove";
  G6Event["CANVAS_MOUSEUP"] = "canvas:mouseup";
  G6Event["CANVAS_CLICK"] = "canvas:click";
  G6Event["CANVAS_MOSUELEAVE"] = "canvas:mouseleave";
  G6Event["CANVAS_DRAGSTART"] = "canvas:dragstart";
  G6Event["CANVAS_DRAG"] = "canvas:drag";
  G6Event["CANVAS_DRAGEND"] = "canvas:dragend"; // combo

  G6Event["COMBO_CLICK"] = "combo:click";
  G6Event["COMBO_CONTEXTMENU"] = "combo:contextmenu";
  G6Event["COMBO_DBLCLICK"] = "combo:dblclick";
  G6Event["COMBO_DRAGSTART"] = "combo:dragstart";
  G6Event["COMBO_DRAG"] = "combo:drag";
  G6Event["COMBO_DRAGEND"] = "combo:dragend";
  G6Event["COMBO_MOUSEENTER"] = "combo:mouseenter";
  G6Event["COMBO_MOUSELEAVE"] = "combo:mouseleave";
  G6Event["COMBO_MOUSEMOVE"] = "combo:mousemove";
  G6Event["COMBO_DROP"] = "combo:drop";
  G6Event["COMBO_DRAGOVER"] = "combo:dragover";
  G6Event["COMBO_DRAGLEAVE"] = "combo:dragleave";
  G6Event["COMBO_DRAGENTER"] = "combo:dragenter";
})(G6Event || (exports.G6Event = G6Event = {}));
