// _@antv_g6@3.5.2@@antv/g6/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Graph", {
  enumerable: true,
  get: function get() {
    return _graph.default;
  }
});
Object.defineProperty(exports, "TreeGraph", {
  enumerable: true,
  get: function get() {
    return _treeGraph.default;
  }
});
Object.defineProperty(exports, "Layout", {
  enumerable: true,
  get: function get() {
    return _layout.default;
  }
});
Object.defineProperty(exports, "Global", {
  enumerable: true,
  get: function get() {
    return _global.default;
  }
});
Object.defineProperty(exports, "Util", {
  enumerable: true,
  get: function get() {
    return _util.default;
  }
});
exports.Algorithm = exports.default = exports.registerBehavior = exports.Menu = exports.Bundling = exports.Grid = exports.Minimap = exports.registerLayout = exports.registerEdge = exports.registerCombo = exports.registerNode = void 0;

var _package = require("../package.json");

var _behavior = _interopRequireDefault(require("./behavior"));

var _graph = _interopRequireDefault(require("./graph/graph"));

var _treeGraph = _interopRequireDefault(require("./graph/tree-graph"));

var _shape = _interopRequireDefault(require("./shape"));

var _layout = _interopRequireDefault(require("./layout"));

var _global = _interopRequireDefault(require("./global"));

var _util = _interopRequireDefault(require("./util"));

var _plugins = _interopRequireDefault(require("./plugins"));

var Algorithm = _interopRequireWildcard(require("./algorithm"));

exports.Algorithm = Algorithm;

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var registerNode = _shape.default.registerNode;
exports.registerNode = registerNode;
var registerEdge = _shape.default.registerEdge;
exports.registerEdge = registerEdge;
var registerCombo = _shape.default.registerCombo;
exports.registerCombo = registerCombo;
var registerBehavior = _behavior.default.registerBehavior;
exports.registerBehavior = registerBehavior;
var registerLayout = _layout.default.registerLayout;
exports.registerLayout = registerLayout;
var Minimap = _plugins.default.Minimap;
exports.Minimap = Minimap;
var Grid = _plugins.default.Grid;
exports.Grid = Grid;
var Bundling = _plugins.default.Bundling;
exports.Bundling = Bundling;
var Menu = _plugins.default.Menu;
exports.Menu = Menu;
var _default = {
  version: _package.version,
  Graph: _graph.default,
  TreeGraph: _treeGraph.default,
  Util: _util.default,
  registerNode: _shape.default.registerNode,
  registerEdge: _shape.default.registerEdge,
  registerCombo: _shape.default.registerCombo,
  registerBehavior: _behavior.default.registerBehavior,
  registerLayout: _layout.default.registerLayout,
  Layout: _layout.default,
  Global: _global.default,
  Minimap: _plugins.default.Minimap,
  Grid: _plugins.default.Grid,
  Bundling: _plugins.default.Bundling,
  Menu: _plugins.default.Menu,
  Algorithm: Algorithm
};
exports.default = _default;
